src/Features/CSharp/Portable/CodeRefactorings/ConvertIfToSwitch/CSharpConvertIfToSwitchCodeRefactoringProvider.Pattern.cs

```diff
@@ -0,0 +1,78 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using Microsoft.CodeAnalysis.CodeRefactorings.ConvertIfToSwitch;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+
+namespace Microsoft.CodeAnalysis.CSharp.CodeRefactorings.ConvertIfToSwitch
+{
+    partial class CSharpConvertIfToSwitchCodeRefactoringProvider
+    {
+        public abstract class Pattern
```

Comment:
Looking at the code, I just used a single method of ISyntaxFactsService (AreEquivalent) so I think that would not make sense to generalize it that way — AreEquivalentCore just turns to an abstract GetSyntaxFactsService which need to be repeatedly called in every use of my AreEquivalent method.

About SyntaxGenerator, I took a glance at implementations. To generalize pattern-matching we will need to extensively rely on casts. I think that's not ok since we have (necessary) type parameters in the base class to avoid just that.